src/vuln_analysis/utils/java_segmenters_with_methods.py

```diff
@@ -0,0 +1,1749 @@
+    annotations and comments so that method-level annotations do NOT appear in the
+    returned snippet. Returns first non-annotation token index.
+    """
+    n = len(src)
+    i = _skip_ws_comments(src, i)
+    while i < n:
+        if src[i] == '@':
+            i += 1
+            while i < n and (src[i].isalnum() or src[i] in ('_', '$', '.')):
+                i += 1
+            i = _skip_ws_comments(src, i)
+            if i < n and src[i] == '(':
+                i = _match_balanced_parens(src, i) + 1
+            i = _skip_ws_comments(src, i)
+            continue
+        j = _skip_ws_comments(src, i)
+        if j != i:
+            i = j
+            continue
+        break
+    return i
+
+def _back_ident(src: str, i: int) -> Tuple[int, Optional[str]]:
+    """
+    From position i (typically the '(' of a parameter list), scan left to find the
+    simple identifier just before it. Returns (start_index, name) or (i, None).
+    """
+    j = i - 1
+    # skip ws
+    while j >= 0 and src[j] in ' \t\r\n\f':
+        j -= 1
+    if j < 0:
+        return i, None
+    end = j + 1
+    while j >= 0 and _is_ident_part(src[j]):
+        j -= 1
+    start = j + 1
+    name = src[start:end]
+    if name and _is_ident_start(name[0]):
+        return start, name
+    return i, None
+
+def _find_left_boundary(src: str, before: int) -> int:
+    """
+    Return index just after the nearest real boundary (';', '{', '}') occurring before `before`,
+    ignoring comments/strings and jumping over balanced () for speed.
+    IMPORTANT: Treat '}', '{', ';' as immediate boundaries BEFORE any jump, so we don't leap
+    across the previous method body and accidentally glue multiple methods together.
+    """
+    if _DELIM_SRC_ID != id(src):
+        # simple forward fallback
+        i = 0
+        last = 0
+        before = max(0, min(before, len(src)))
+        while i < before:
+            j = _skip_ws_comments(src, i)
+            if j != i:
+                i = j; continue
+            if i < before and (src.startswith('"""', i) or src[i] in ('"', "'")):
+                i = _skip_java_string_like(src, i); continue
+            if src[i] in (';', '{', '}'):
+                last = i + 1
+            i += 1
+        return last
+
+    # fast backward using indexes
+    i = min(before - 1, len(src) - 1)
+    while i >= 0:
+        # hop out of ignored span
+        span = _ignored_span_at(i)
+        if span:
+            i = span[0] - 1
+            continue
+
+        ch = src[i]
+
+        # FIRST: boundary check
+        if ch in (';', '{', '}'):
+            return i + 1
+
+        # jump across groups for speed (do NOT jump across '}' – handled above)
+        if ch == ')':
+            o = _PAR_C2O.get(i) if _DELIM_SRC_ID == id(src) else None
+            i = (o - 1) if o is not None else (i - 1)
+            continue
+
+        i -= 1
+    return 0
+
+def _match_paren_backwards(src: str, i: int) -> int:
```

Comment:
@tmihalac Function intention and purpose is not quite clear:
 please document function purpose succinctly , what it gets as input, what it does, and what is the integer that it returns to callers